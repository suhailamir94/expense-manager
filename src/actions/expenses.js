export const addExpense = (
  { description = "", note = "", amount = 0, createdAt = 0 },
  userId
) => ({
  type: "ADD_EXPENSE",
  expense: {
    description,
    note,
    amount,
    createdAt,
  },
  userId,
});

//EDIT_EXPENSE
export const removeExpense = ({ id, userId }) => ({
  type: "REMOVE_EXPENSE",
  id,
  userId,
});

//REMOVE_EXPENSE
export const editExpense = (id, updates, userId) => ({
  type: "EDIT_EXPENSE",
  id,
  updates,
  userId,
});
