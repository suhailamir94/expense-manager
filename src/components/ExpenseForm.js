import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";

const now = moment();

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.onDescriptionChange = this.onDescriptionChange.bind(this);
    this.onAmountChange = this.onAmountChange.bind(this);
    this.onNoteChange = this.onNoteChange.bind(this);
    this.onDateChange = this.onDateChange.bind(this);
    this.onFocusChange = this.onFocusChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
    this.state = {
      description: props.expense ? props.expense.description : "",
      amount: props.expense ? props.expense.amount.toString() : "",
      note: props.expense ? props.expense.note : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calenderFocused: false,
      error: "",
    };
  }
  onDescriptionChange(e) {
    const description = e.target.value;
    this.setState(prevState => ({ ...prevState, description }));
  }
  onNoteChange(e) {
    const note = e.target.value;
    this.setState(prevState => ({ ...prevState, note }));
  }

  onAmountChange(e) {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}$/))
      this.setState(prevState => ({ ...prevState, amount }));
  }

  onDateChange(createdAt) {
    if (createdAt)
      this.setState(() => ({
        createdAt,
      }));
  }

  onFocusChange({ focused }) {
    this.setState(() => ({ calenderFocused: focused }));
  }
  onSubmit(e) {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: "Please provide description and amount" }));
    } else {
      this.state.error && this.setState(() => ({ error: "" }));
      this.props.onSubmit({
        description: this.state.description,
        amount: this.state.amount,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note,
      });
    }
  }

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Descreption"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt} // momentPropTypes.momentObj or null
            onDateChange={this.onDateChange} // PropTypes.func.isRequired
            focused={this.state.calenderFocused} // PropTypes.bool
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="Add a note for for your expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          ></textarea>
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}
