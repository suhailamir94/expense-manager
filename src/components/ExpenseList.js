import React from "react";
import { connect } from "react-redux";

import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";

export const ExpenseList = props => {
  return (
    <div>
      <h2>Expense List</h2>
      {props.expenses.length ? (
        props.expenses.map(expense => (
          <ExpenseListItem key={expense.id} {...expense} />
        ))
      ) : (
        <p>No Expenses Found!!</p>
      )}
    </div>
  );
};
const mapStateToProps = state => {
  return {
    expenses: selectExpenses(state.expenses, state.filters),
  };
};
export default connect(mapStateToProps)(ExpenseList);
