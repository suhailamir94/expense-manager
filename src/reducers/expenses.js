const expensesReducerDefaultState = [];
const expensesReducer = (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case "GET_EXPENSES_SUCCEEDED":
      return [...action.expenses];
    case "ADD_EXPENSE_SUCCEEDED":
      return [...state, action.expense];
    case "REMOVE_EXPENSE_SUCCEEDED":
      return state.filter(e => e.id != action.id);
    case "EDIT_EXPENSE_SUCCEEDED":
      return state.map(expense =>
        expense.id === action.id ? { ...expense, ...action.updates } : expense
      );
    case "ADD_EXPENSE_FAILED":
      return state;
    case "GET_EXPENSES_FAILED":
      return state;
    case "EDIT_EXPENSE_FAILED":
      return state;
    case "REMOVE_EXPENSE_FAILED":
      return state;
    default:
      return state;
  }
};

export default expensesReducer;
