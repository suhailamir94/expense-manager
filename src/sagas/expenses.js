import { call, put, takeEvery, takeLatest } from "redux-saga/effects";
import {
  addExpenseFirebase,
  getAllExpensesFirebase,
  editExpenseFirebase,
  removeExpenseFirebase,
} from "../firebase/firebase";

// worker Saga: will be fired on USER_FETCH_REQUESTED actions
function* addExpenseAsync(action) {
  try {
    const id = yield call(addExpenseFirebase, [action.expense, action.userId]);
    yield put({
      type: "ADD_EXPENSE_SUCCEEDED",
      expense: { id, ...action.expense },
    });
  } catch (e) {
    yield put({ type: "ADD_EXPENSE_FAILED", message: e.message });
  }
}
function* editExpenseAsync({ id, updates, userId }) {
  try {
    yield call(editExpenseFirebase, [id, updates, userId]);
    yield put({
      type: "EDIT_EXPENSE_SUCCEEDED",
      id,
      updates,
    });
  } catch (e) {
    yield put({ type: "ADD_EXPENSE_FAILED", message: e.message });
  }
}
function* removeExpenseAsync({ id, userId }) {
  try {
    yield call(removeExpenseFirebase, [id, userId]);
    yield put({
      type: "REMOVE_EXPENSE_SUCCEEDED",
      id,
    });
  } catch (e) {
    yield put({ type: "REMOVE_EXPENSE_FAILED", message: e.message });
  }
}
function* getExpensesAsync(action) {
  try {
    const expenses = yield call(getAllExpensesFirebase, action.userId);
    yield put({
      type: "GET_EXPENSES_SUCCEEDED",
      expenses,
    });
  } catch (e) {
    yield put({ type: "GET_EXPENSES_FAILED", message: e.message });
  }
}
/*
  Starts fetchUser on each dispatched `USER_FETCH_REQUESTED` action.
  Allows concurrent fetches of user.
*/

export function* getAllExpensesSaga() {
  yield takeEvery("GET_EXPENSES", getExpensesAsync);
}

export function* addExpenseSaga() {
  yield takeEvery("ADD_EXPENSE", addExpenseAsync);
}

export function* editExpenseSaga() {
  yield takeEvery("EDIT_EXPENSE", editExpenseAsync);
}

export function* removeExpenseSaga() {
  yield takeEvery("REMOVE_EXPENSE", removeExpenseAsync);
}
