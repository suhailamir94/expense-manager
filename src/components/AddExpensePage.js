import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { addExpense } from "../actions/expenses";

export const AddExpensePage = props => (
  <div>
    <h1>Add Expense</h1>
    <ExpenseForm
      onSubmit={expense => {
        props.startAddExpense(expense); // this onSubmit is passed to props by the mapDispatchToProps method defined below
        // this is how we redirect from one page to another through JS in React Router
        props.history.push("/dashboard");
      }}
    />
  </div>
);

const mapDispatchToProps = dispatch => ({
  startAddExpense: expense => dispatch(addExpense(expense)),
});
export default connect(undefined, mapDispatchToProps)(AddExpensePage);
