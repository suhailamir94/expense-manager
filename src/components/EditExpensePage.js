import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { editExpense, removeExpense } from "../actions/expenses";

export const EditExpensePage = props => {
  return (
    <div>
      <ExpenseForm
        expense={props.expense}
        onSubmit={expense => {
          props.onSubmit(props.expense.id, expense);
          // this is how we redirect from one page to another through JS in React Router
          props.history.push("/");
        }}
      />
      <button
        onClick={() => {
          props.onClickRemove(props.expense.id);
          // props.dispatch(removeExpense({ id: props.expense.id }));
          props.history.push("/");
        }}
      >
        REMOVE
      </button>
    </div>
  );
};

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id),
});

const mapDispatchToProps = dispatch => ({
  onSubmit: (id, expense) => dispatch(editExpense(id, expense)),
  onClickRemove: id => dispatch(removeExpense({ id })),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
