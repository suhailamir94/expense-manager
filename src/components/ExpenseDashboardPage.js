import React from "react";
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";
import ExpensesSummary from "./ExpensesSummary";

const ExpenseDashboardPage = () => (
  <div className="container-column">
    <div className="container">
      <ExpenseListFilters />
      <ExpensesSummary />
      <ExpenseList />
    </div>
  </div>
);
export default ExpenseDashboardPage;
