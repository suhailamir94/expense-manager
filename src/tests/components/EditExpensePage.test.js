import React from "react";
import { shallow } from "enzyme";
import { EditExpensePage } from "../../components/EditExpensePage";
import expenses from "../fixtures/expenses";

let onSubmitSpy, onClickRemoveSpy, history, wrapper;

beforeEach(() => {
  onSubmitSpy = jest.fn();
  onClickRemoveSpy = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditExpensePage
      onSubmit={onSubmitSpy}
      history={history}
      onClickRemove={onClickRemoveSpy}
      expense={expenses[0]}
    />
  );
});

test("should render EditExpensepage correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should handle onSubmit", () => {
  wrapper.find("ExpenseForm").prop("onSubmit")(expenses[0]);
  expect(history.push).toHaveBeenLastCalledWith("/");
  expect(onSubmitSpy).toHaveBeenLastCalledWith(expenses[0].id, expenses[0]);
});

test("should handle onClickRemove", () => {
  wrapper.find("button").prop("onClick")(expenses[0]);
  expect(history.push).toHaveBeenLastCalledWith("/");
  expect(onClickRemoveSpy).toHaveBeenLastCalledWith(expenses[0].id);
});
