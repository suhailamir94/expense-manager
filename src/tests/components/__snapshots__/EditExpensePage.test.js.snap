// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensepage correctly 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 195,
        "createdAt": 0,
        "description": "gum",
        "id": "1",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    REMOVE
  </button>
</div>
`;
